sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -1377,17 +1394,34 @@ case class StringInstr(str: Expression, substr: Expression)
   override def left: Expression = str
   override def right: Expression = substr
   override def dataType: DataType = IntegerType
-  override def inputTypes: Seq[DataType] = Seq(StringType, StringType)
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(StringTypeAnyCollation, StringTypeAnyCollation)
 
   override def nullSafeEval(string: Any, sub: Any): Any = {
-    string.asInstanceOf[UTF8String].indexOf(sub.asInstanceOf[UTF8String], 0) + 1
+    val collationId = left.dataType.asInstanceOf[StringType].collationId
+    string.asInstanceOf[UTF8String].indexOf(sub.asInstanceOf[UTF8String], 0, collationId) + 1
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      return defaultCheck
+    }
+
+    val collationId = left.dataType.asInstanceOf[StringType].collationId
+    CollationTypeConstraints.checkCollationCompatibility(collationId, children.map(_.dataType))
```

Comment:
@dbatomic I think we should have a framework-level collation compatibility check.

Reply:
+1 I believe this check should be removed after merging implicit casting. One problem with the check is that in future it does not take into account the fact that we can have indeterminate collation and blindly calls fetchCollation which can fail to return a collator. This PR https://github.com/apache/spark/pull/45383 will remove it, as implicit casting should make sure all collations are the same and that they are not indeterminate if that is needed.

Reply:
Btw, let's wait for implicit casting before merging this PR. Implicit casting should change things both in constraints space and also how we write tests.